src/effects/VST3/VST3Effect.cpp

```diff
@@ -0,0 +1,787 @@
+}
+
+bool VST3Effect::IsInteractive()
+{
+   return mEditController != nullptr;
+}
+
+bool VST3Effect::IsDefault()
+{
+   return false;
+}
+
+bool VST3Effect::IsLegacy()
+{
+   return false;
+}
+
+bool VST3Effect::SupportsRealtime()
+{
+   return true;
+}
+
+bool VST3Effect::SupportsAutomation()
+{
+   if(mEditController == nullptr)
+      return false;
+
+   using namespace Steinberg;
+   
+   for(int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+   {
+      Vst::ParameterInfo parameterInfo;
+      if(mEditController->getParameterInfo(i, parameterInfo) == kResultOk)
+      {
+         if(parameterInfo.flags & Vst::ParameterInfo::kCanAutomate)
+            return true;
+      }
+   }
+
+   return false;
+}
+
+bool VST3Effect::GetAutomationParameters(CommandParameters& parms)
+{
+   if(mEditController == nullptr)
+      return false;
+
+   using namespace Steinberg;
+   
+   for(int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+   {
+      Vst::ParameterInfo parameterInfo;
+      if(mEditController->getParameterInfo(i, parameterInfo) == kResultOk)
+      {
+         parms.Write(
+            std::to_string(parameterInfo.id),
+            mEditController->getParamNormalized(parameterInfo.id)
+         );
+      }
+   }
+
+   return true;
+}
+
+bool VST3Effect::SetAutomationParameters(CommandParameters& parms)
+{
+   using namespace Steinberg;
+
+   if(mComponentHandler== nullptr)
+      return false;
+
+   for(int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+   {
+      Vst::ParameterInfo parameterInfo;
+      if(mEditController->getParameterInfo(i, parameterInfo) != kResultOk)
+         continue;
+
+      Vst::ParamValue value;
+      if(parms.Read(std::to_string(parameterInfo.id), &value))
+      {
+         mComponentHandler->beginEdit(parameterInfo.id);
+         mComponentHandler->performEdit(parameterInfo.id, value);
+         mComponentHandler->endEdit(parameterInfo.id);
+         mEditController->setParamNormalized(parameterInfo.id, value);
+      }
+   }
+
+   return true;
+}
+
+bool VST3Effect::LoadUserPreset(const RegistryPath& name)
+{
+   //Dummy
+   return true;
+}
+
+bool VST3Effect::SaveUserPreset(const RegistryPath& name)
+{
+   //Dummy
+   return true;
+}
+
+RegistryPaths VST3Effect::GetFactoryPresets()
+{
+   return { };
+}
+
+bool VST3Effect::LoadFactoryPreset(int id)
+{
+   return true;
+}
+
+bool VST3Effect::LoadFactoryDefaults()
+{
+   using namespace Steinberg;
+
+   if(mEditController != nullptr)
+   {
+      for(int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+      {
+         Vst::ParameterInfo parameterInfo;
+         if(mEditController->getParameterInfo(i, parameterInfo) == kResultOk)
+         {
+            if(parameterInfo.flags & Vst::ParameterInfo::kIsReadOnly)
+               continue;
+
+            mComponentHandler->performEdit(parameterInfo.id, parameterInfo.defaultNormalizedValue);
```

Comment:
This isn't null checked, but I think you guarantee the invariant that it is not null, if mEditController is not null